### DISCUSSION TITLE
Menu of actions for text selection

### DISCUSSION THREAD

**Discussion by amitbha** [orbitalquark/textadept#394] [Ideas]
![highlight-on-kindle-for-pc](https://user-images.githubusercontent.com/4060625/232002547-1b53f200-4ea8-475d-b25f-75c77e3885bb.png)
Just like the kindle reading app, I want to select the text, then popup a menu to action further, like search, translate, etc. More handy than keyboard shortcuts.
The user list seems doable, but I can't find the event for selection-complete.
Is there a way to do this?

**Comment by rgieseke** [comment]
This looks like a similar question to yours: https://github.com/jacobslusser/ScintillaNET/issues/149

(Textadept uses Scintilla.)

**Reply by amitbha** [reply]
> This looks like a similar question to yours: [jacobslusser/ScintillaNET#149](https://github.com/jacobslusser/ScintillaNET/issues/149)

```lua
events.connect(events.UPDATE_UI, function(updated)
    if updated & buffer.UPDATE_SELECTION > 0 then
        ui.dialogs.message('selected')
    end
end)
```
TA get strucked when startup...

**Reply by rgieseke** [reply]
Maybe the dialog is blocking this, try it like this?

```lua
events.connect(events.UPDATE_UI, function(updated)
    if updated & buffer.UPDATE_SELECTION > 0 then
        ui.statusbar_text = 'selected'
    end
end)
```

And then check for shift being pressed, now it fires at every cursor move.

**Reply by rgieseke** [reply]
Or you could connect your event wrapped in an `events.INITIALIZED` event so it's not firing during Textadept's startup.

**Comment by orbitalquark** [comment]
Something like this might work in your *~/.textadept/init.lua*:

```lua
table.insert(textadept.menu.context_menu, {
  'Translate', function() ui.print('translate ', buffer:get_sel_text()) end
})

events.connect(events.UPDATE_UI, function(updated)
  if updated & buffer.UPDATE_SELECTION == 0 or buffer.selection_empty then return end
  ui.popup_menu(ui.context_menu)
end)
```

**Comment by orbitalquark** [comment]
If you want to use a user list, then you'll want to connect to `events.USER_LIST_SELECTION` (https://orbitalquark.github.io/textadept/api.html#events.USER_LIST_SELECTION).

**Comment by amitbha** [comment]
Many thanks.
The left problems:
1. don't know when the selection stop
2. sometimes conflict with TAB when indent the selections

~~~lua
view.auto_c_max_width = 4
events.connect(events.UPDATE_UI, function(updated)
    if updated & buffer.UPDATE_SELECTION == 0 or buffer.selection_empty then return end
    -- TODO: popup user list when the selection stop, not at the beginning
    if not buffer:auto_c_active() then buffer:user_list_show(10000, 'translate search') end
end)

events.connect(events.USER_LIST_SELECTION, function(id, text, pos)
    if id == 10000 then
        if text == 'search' then ui.dialogs.message{title = text} end
    end
    buffer:cancel()
end)
~~~

**Reply by orbitalquark** [reply]
Yeah, I see now that what I suggested isn't very robust. I'm not sure how to do what you are trying to do. I would just stick to having a key binding show the user list or context menu with a translate option or any others you want to add after making a selection.

**Reply by amitbha** [reply]
I usually use TA as light-weighted reading & editing tool for markdown files, while the mouse is used most.
And sometimes design and remember key bindings is burdensome, like shortcuts-conflicts, consistency with other editors.

**Reply by jxzwp** [reply]
Updated 10:21 UTC, tabbing now works, I think... 

There is no selection "stop" event but a user list could be shown when the selection hasn't changed for a period of time by using a timeout. So it would be like "dwelling" on a selection. I've tried the implement this below. 
Some Notes:
+ `dwell_sel_time` controls how long to wait before showing the list, currently half a second. Change it to your preference.
+ Tabbing to indent when the list is showing still works.
+ Pressing any modifier key closes the list so copy and paste etc aren't hindered.
+ I've added the keybinding Ctrl+5 to show the list on demand, change this to your preference.
+ This was tested on a recent Textadept 12 nightly on Linux (curses version not tested)
+ @orbitalquark There's a typo in the api docs for `events.AUTO_C_CANCELLED` is which is shown as `events.AUTO_C_CANCELED`.

Put the code below in your ~/.textadept/init.lua to try it.
```lua
buffer.auto_c_max_height = 4  -- max entries to show in list before scrolling

local dwell_sel_time = 0.5  -- seconds that the selection must remain unchanged before showing user list
local dwell_sel_start, dwell_sel_end = 0, 0  -- track start/end of selection

local function show_list(s, e)
  if s and (s ~= dwell_sel_start or e ~= dwell_sel_end) then return end
  if buffer.selection_empty then return end
  if not buffer:auto_c_active() then 
    buffer:user_list_show(10000, 'search translate') 
  end 
end

-- Optional keybinding to show list on demand
keys['ctrl+5'] = show_list

events.connect(events.UPDATE_UI, function(updated) 
  if (updated & buffer.UPDATE_SELECTION == 0) or buffer.selection_empty then return end 
  dwell_sel_start = buffer.selection_start
  dwell_sel_end = buffer.selection_end
  timeout(dwell_sel_time, show_list, buffer.selection_start, buffer.selection_end)
end) 

-- Track the last keypress, so we know when Tab has been pressed
local last_key 
events.connect(events.KEYPRESS, function(key) last_key = key end)

-- When the user presses Tab with the user list showing, cancel the list 
-- then press Tab again programmatically, so indenting still works.
-- If Enter is pressed with the list showing, select the list item as normal
events.connect(events.USER_LIST_SELECTION, function(id, text, pos) 
  if last_key == '\t' and not buffer.selection_empty then 
    buffer:auto_c_cancel()
    buffer:tab() 
    return false  -- stop event propagating
  end  
  if id == 10000 then
    if text == 'search' then 
      ui.dialogs.message{title = text}
      -- Do stuff with buffer:get_sel_text() here
    end 
  end 
  buffer:auto_c_cancel() 
end, 1)

-- When any modifier key is pressed cancel the user list if it's shown.
-- This is so the user list doesn't hinder cut and paste and other functionality
-- CAUTION: this relies on the undocumented `events.KEY` event
-- and borrows code from core/keys.lua:130 (Thanks Mitchell)
local MOD_SHIFT, MOD_CTRL, MOD_ALT, MOD_META = _SCINTILLA.constants.MOD_SHIFT,
  _SCINTILLA.constants.MOD_CTRL, _SCINTILLA.constants.MOD_ALT, _SCINTILLA.constants.MOD_META

events.connect(events.KEY, function(code, modifiers) 
  local shift, ctrl, alt, cmd = modifiers & MOD_SHIFT > 0, modifiers & MOD_CTRL > 0,
    modifiers & MOD_ALT > 0, modifiers & MOD_META > 0
  if shift or ctrl or alt or cmd then buffer:auto_c_cancel() end
end)
```

**Reply by orbitalquark** [reply]
>     * @orbitalquark There's a typo in the api docs for `events.AUTO_C_CANCELLED` is which is shown as `events.AUTO_C_CANCELED`.

It's not a typo. There was a bug in the API generation from Scintilla that has been fixed via https://github.com/orbitalquark/textadept/commit/4018e7fd6874b97a91976a5cffa3dc1d058ac11b and will be in the next nightly release. Thanks for catching and reporting the error.
